src/Caching/Hybrid/src/Internal/RecyclableArrayBufferWriter.cs

```diff
@@ -0,0 +1,184 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Extensions.Caching.Hybrid.Internal;
+
+// this is effectively a cut-down re-implementation of ArrayBufferWriter
+// from https://github.com/dotnet/runtime/blob/6cd9bf1937c3b4d2f7304a6c534aacde58a202b6/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs
+// except it uses the array pool for allocations
+internal sealed class RecyclableArrayBufferWriter<T> : IBufferWriter<T>, IDisposable
```

Comment:
Why disposable?  The callers seem to `Dispose` manually in order to avoid doing so on exceptions.

Reply:
Utility code, but yes in these scenarios we could choose not to retain that. Semantically, it is valid, though

Reply:
I'm not sure what you mean when you say that it's semantically valid.  It sounds like, in most cases, it should _not_ be used with `using` statements, meaning that it doesn't behave like a normal `IDisposable`?

Reply:
In most cases it probably *should*; we're in an unusual situation here, because we're in the middle of async calls involving leased buffers, which makes things risky. I'll add comments.

Reply:
https://github.com/dotnet/aspnetcore/pull/55251